### PR TITLE
Adds the Alternate Sprites subsystem to the game, with an example case.

### PR DESCRIPTION
**Pull request by Iamgoofball**
:cl: Iamgoofball
rscadd: Security uniforms are now subcontracted to other companies.
/:cl:

The Alternate Sprites subsystem allows for setting the round's sprite theme for certain items in groups. In this case, it is used to subcontract out the security uniform design to various companies.

### CODE REVIEW

**Comment by Iamgoofball** [comment]
Also, this is Cheridan's birthday PR.

Happy birthday @Cheridan

**Comment by Bawhoppen** [comment]
>goof thumbs-upping his own PR

File: code/controllers/subsystem/alt_sprites.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	A.icon_state = initial(chosen.icon_state)
+	A.item_state = initial(chosen.item_state)
+	A.item_color = initial(chosen.item_color)
+	A.update_icon()
+
+/datum/sprite_alt
+	var/name = "http://www."
+	var/icon_state = "porn"
+	var/item_state = "hub"
+	var/item_color = ".com"
```
Comment (MMMiracles) [inline code/controllers/subsystem/alt_sprites.dm]: didnt you get yelled at last time for these memes

Comment (JJRcop) [inline code/controllers/subsystem/alt_sprites.dm]: He most definitely did

Comment (RemieRichards) [inline code/controllers/subsystem/alt_sprites.dm]: @Iamgoofball 8 hours to fix this or this PR dies.

File: code/controllers/subsystem/alt_sprites.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	var/help_me = pick(sec_themes)
+	var/datum/sprite_theme/S = new help_me
+	for(var/O in S.overrides)
+		sprite_overrides += new O
+	for(var/datum/sprite_override/OR in sprite_overrides)
+		OR.adjust_all_sprites()
+
+
+/datum/sprite_theme
+	var/name = "redtube"
```
Comment (RemieRichards) [inline code/controllers/subsystem/alt_sprites.dm]: same as the other comment.

File: code/controllers/subsystem/alt_sprites.dm [hunk 3]
```diff
@@ -0,0 +1,57 @@
+var/datum/subsystem/altsprites/SSaltsprites
+
+/datum/subsystem/altsprites
+	name = "Alternate Sprites"
+	init_order = -420
```
Comment (RemieRichards) [inline code/controllers/subsystem/alt_sprites.dm]: use a real value.

File: code/controllers/subsystem/alt_sprites.dm [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/datum/sprite_override/proc/adjust_all_sprites()
+	for(var/obj/item/A in list_to_search)
+		if(A.type != item_type)
+			continue
+		A.icon_state = initial(chosen.icon_state)
+		A.item_state = initial(chosen.item_state)
+		A.item_color = initial(chosen.item_color)
+		A.update_icon()
+
+/datum/sprite_override/proc/adjust_a_sprite(var/obj/item/A)
```
Comment (RemieRichards) [inline code/controllers/subsystem/alt_sprites.dm]: no ```var/``` in proc arguments

File: code/game/objects/items.dm [hunk 5]
```diff
@@ -108,6 +108,10 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
 	..()
 	for(var/path in actions_types)
 		new path(src)
+	if(SSaltsprites)
+		for(var/datum/sprite_override/O in SSaltsprites.sprite_overrides)
```
Comment (RemieRichards) [inline code/game/objects/items.dm]: this should be a typecache.

Comment (Iamgoofball) [inline code/game/objects/items.dm]: What the fuck is a typecache

Comment (RemieRichards) [inline code/game/objects/items.dm]: a vastly faster way to do an "istype in list" style situation.

File: code/controllers/subsystem/alt_sprites.dm (hunk 1 above)
Comment (JJRcop) [inline code/controllers/subsystem/alt_sprites.dm]: Iamgoofball#5

**Comment by TechnoAlchemisto** [comment]
plz no redsec

**Comment by TechnoAlchemisto** [comment]
that is if redsec is old red sec

File: code/game/objects/items.dm (hunk 5 above)
Comment (JJRcop) [inline code/game/objects/items.dm]: @RemieRichards You didn't explain what it was and how to make one

**Comment by kevinz000** [comment]
THE NEW SPRITES DISGUST US.

**Comment by Iamgoofball** [comment]
Is this test merged?

**Comment by Boredone** [comment]
So, the sprite changes seem to be affecting the children of armor/vest which include the warden's jacket and the Cappy's Carapace and Parade Outfit. So should probobly add a check to keep children from being effected.

**Comment by Iamgoofball** [comment]
I'll check it out, there's apparently some odd behavior with byond RNG

**Comment by Bawhoppen** [comment]
you can tell goof was trying to actually get this merged by being half-professional by actually naming the PR something meaningful

**Comment by Iamgoofball** [comment]
Warden jackets have their own override btw

**Comment by PKPenguin321** [comment]
https://github.com/tgstation/tgstation/issues/19874

File: code/controllers/subsystem/alt_sprites.dm [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+/datum/subsystem/altsprites
+	name = "Alternate Sprites"
+	init_order = -420
+	flags = SS_NO_FIRE
+	var/list/sprite_overrides = list()
+
+/datum/subsystem/altsprites/New()
+	NEW_SS_GLOBAL(SSaltsprites)
+	var/list/sec_themes = list(/datum/sprite_theme/oldsec, /datum/sprite_theme/greysec, /datum/sprite_theme/copsec, /datum/sprite_theme/corpsec, /datum/sprite_theme/redsec)
+	var/help_me = pick(sec_themes)
```
Comment (Iamgoofball) [inline code/controllers/subsystem/alt_sprites.dm]: so uh, apparently this stopped working

any clues

**Comment by KorPhaeron** [comment]
the blue ones are hideous

File: code/game/objects/items.dm (hunk 5 above)
Comment (optimumtact) [inline code/game/objects/items.dm]: goof should know how to do these things

Comment (Iamgoofball) [inline code/game/objects/items.dm]: Type cache is new to me actually. I was gonna ask about it when I get home, unless it's initial meme

**Comment by YakumoChen** [comment]
Alternate borg sprites please, so we can port the ones from vg

**Comment by KorPhaeron** [comment]
>it's a "goof opens an unfinished PR and lets it sit for a year" episode

**Comment by Iamgoofball** [comment]
i cant figure out how to implement the typecache memes as apparently im not allowed to use the existing procs for it

**Comment by TechnoAlchemisto** [comment]
You should probably ask Remie about it

**Comment by RemieRichards** [comment]
It's just an associative list/dict (which is a BASIC tool in many languages) where the keys are types and the values are something truth-y, as I said before your values should be the actual datums.

**Comment by optimumtact** [comment]
stale

**Pull request closed without merging**